File: packages/typescript/src/inboundCanonical.ts
/** Mirrors `agentInboundWebhook.signing.ts` in mcp-main. */

export type InboundWebhookBody = Record<string, unknown> & {
  text?: string;
  idempotencyKey?: string;
};

type CanonicalValue =
  | string
  | number
  | boolean
  | null
  | CanonicalValue[]
  | { [key: string]: CanonicalValue };

const STRUCTURED_SIGNING_KEYS = [
  'action',
  'amount',
  'chain',
  'changes',
  'clientTimestamp',
  'confidence',
  'context',
  'currentPnlPct',
  'entryPrice',
  'entryZone',
  'eventType',
  'exitPrice',
  'exitReason',
  'expiresAt',
  'ideaId',
  'instrument',
  'leverageX',
  'metadata',
  'orderType',
  'pnlPct',
  'positionRef',
  'reasoning',
  'relatedTradeId',
  'riskMgmt',
  'sellSizePct',
  'side',
  'symbol',
  'tokenAddress',
  'tradeType',
  'triggers',
  'urgency',
  'version',
] as const;

function canonicalizeValue(value: unknown): CanonicalValue | undefined {
  if (value === undefined) return undefined;
  if (value === null) return null;
  if (typeof value === 'string') return value;
  if (typeof value === 'number') return Number.isFinite(value) ? value : undefined;
  if (typeof value === 'boolean') return value;
  if (Array.isArray(value)) {
    return value
      .map((item) => canonicalizeValue(item))
      .filter((item): item is CanonicalValue => item !== undefined);
  }
  if (typeof value === 'object') {
    const out: { [key: string]: CanonicalValue } = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      const child = canonicalizeValue((value as Record<string, unknown>)[key]);
      if (child !== undefined) out[key] = child;
    }
    return out;
  }
  return undefined;
}

export function inboundWebhookCanonicalJson(body: InboundWebhookBody): string {
  const sorted: Record<string, CanonicalValue> = {};

  if (body.text !== undefined) {
    sorted.text = body.text;
  }

  const idk =
    typeof body.idempotencyKey === 'string' ? body.idempotencyKey.trim() : '';
  if (idk) {
    sorted.idempotencyKey = idk;
  }

  for (const key of STRUCTURED_SIGNING_KEYS) {
    const value = canonicalizeValue(body[key]);
    if (value !== undefined) {
      sorted[key] = value;
    }
  }

  const keys = Object.keys(sorted).sort();
  const obj: Record<string, CanonicalValue> = {};
  for (const k of keys) {
    obj[k] = sorted[k];
  }
  return JSON.stringify(obj);
}

File: packages/rust/src/inbound_canonical.rs
use serde_json::{Map, Value};

const STRUCTURED_SIGNING_KEYS: &[&str] = &[
    "action", "amount", "chain", "changes", "clientTimestamp", "confidence", "context",
    "currentPnlPct", "entryPrice", "entryZone", "eventType", "exitPrice", "exitReason",
    "expiresAt", "ideaId", "instrument", "leverageX", "metadata", "orderType", "pnlPct",
    "positionRef", "reasoning", "relatedTradeId", "riskMgmt", "sellSizePct", "side",
    "symbol", "tokenAddress", "tradeType", "triggers", "urgency", "version",
];

fn canonicalize_value(value: &Value) -> Option<Value> {
    match value {
        Value::Null => Some(Value::Null),
        Value::Bool(_) | Value::String(_) => Some(value.clone()),
        Value::Number(number) => number
            .as_f64()
            .filter(|n| n.is_finite())
            .map(|_| Value::Number(number.clone())),
        Value::Array(items) => Some(Value::Array(
            items
                .iter()
                .filter_map(canonicalize_value)
                .collect::<Vec<_>>(),
        )),
        Value::Object(map) => {
            let mut keys: Vec<_> = map.keys().cloned().collect();
            keys.sort();
            let mut out = Map::new();
            for key in keys {
                if let Some(child) = canonicalize_value(&map[&key]) {
                    out.insert(key, child);
                } else if map[&key].is_null() {
                    out.insert(key, Value::Null);
                }
            }
            Some(Value::Object(out))
        }
    }
}

pub fn inbound_webhook_canonical_json(body: &Map<String, Value>) -> String {
    let mut sorted = Map::new();

    if let Some(Value::String(text)) = body.get("text") {
        sorted.insert("text".to_string(), Value::String(text.clone()));
    }

    if let Some(Value::String(idk)) = body.get("idempotencyKey") {
        let trimmed = idk.trim();
        if !trimmed.is_empty() {
            sorted.insert(
                "idempotencyKey".to_string(),
                Value::String(trimmed.to_string()),
            );
        }
    }

    for key in STRUCTURED_SIGNING_KEYS {
        if let Some(value) = body.get(*key) {
            if let Some(child) = canonicalize_value(value) {
                sorted.insert(key.to_string(), child);
            } else if value.is_null() {
                sorted.insert(key.to_string(), Value::Null);
            }
        }
    }

    let mut keys: Vec<_> = sorted.keys().cloned().collect();
    keys.sort();
    let mut obj = Map::new();
    for key in keys {
        obj.insert(key.clone(), sorted[&key].clone());
    }
    serde_json::to_string(&Value::Object(obj)).unwrap_or_else(|_| "{}".to_string())
}
